src/nginx/t/api_basepath_hard_match.t

```diff
@@ -0,0 +1,221 @@
+control {
+  environment: "http://127.0.0.1:${ServiceControlPort}"
+}
+EOF
+
+$t->write_file('server.pb.txt', <<"EOF");
+service_control_config {
+  check_aggregator_config {
+    cache_entries: 0
+    flush_interval_ms: 100
+  }
+  report_aggregator_config {
+    cache_entries: 10
+    flush_interval_ms: 100
+  }
+}
+api_service_config {
+  base_path: "/api"
+  base_path_hard_match: true
+}
+EOF
+
+
+ApiManager::write_file_expand($t, 'nginx.conf', <<"EOF");
+%%TEST_GLOBALS%%
+daemon off;
+events {
+  worker_connections 32;
+}
+http {
+  %%TEST_GLOBALS_HTTP%%
+  server_tokens off;
+  server {
+    listen 127.0.0.1:${NginxPort};
+    server_name localhost;
+    location / {
+      endpoints {
+        api service.pb.txt;
+        %%TEST_CONFIG%%
+        server_config server.pb.txt;
+        on;
+      }
+      proxy_pass http://127.0.0.1:${BackendPort};
+    }
+  }
+}
+EOF
+
+my $report_done = 'report_done';
+
+$t->run_daemon(\&bookstore, $t, $BackendPort, 'bookstore.log');
+$t->run_daemon(\&servicecontrol, $t, $ServiceControlPort, 'servicecontrol.log', $report_done);
+is($t->waitforsocket("127.0.0.1:${BackendPort}"), 1, 'Bookstore socket ready.');
+is($t->waitforsocket("127.0.0.1:${ServiceControlPort}"), 1, 'Service control socket ready.');
+$t->run();
+
+################################################################################
+
+my $response_first = ApiManager::http_get($NginxPort,'/api/shelves?key=this-is-an-api-key');
+my $response_second = ApiManager::http_get($NginxPort,'/shelves?key=this-is-an-api-key');
+
+is($t->waitforfile("$t->{_testdir}/${report_done}"), 1, 'Report body file ready.');
+
+
+$t->stop_daemons();
+
+my ($response_headers, $response_body) = split /\r\n\r\n/, $response_first, 2;
+
+like($response_headers, qr/HTTP\/1\.1 200 OK/, 'Returned HTTP 200.');
+is($response_body, <<'EOF', 'Shelves returned in the response body.');
+{ "shelves": [
+    { "name": "shelves/1", "theme": "Fiction" },
+    { "name": "shelves/2", "theme": "Fantasy" }
+  ]
+}
+EOF
+
+($response_headers, $response_body) = split /\r\n\r\n/, $response_second, 2;
+
+like($response_headers, qr/HTTP\/1\.1 404 Not Found/, 'Returned HTTP 404.');
+is($response_body, <<'EOF', 'Shelves returned in the response body.');
+{
+ "code": 5,
+ "message": "NOT_FOUND",
+ "details": [
+  {
+   "@type": "type.googleapis.com/google.rpc.DebugInfo",
+   "stackEntries": [],
+   "detail": "application"
+  }
+ ]
+}
+EOF
+
+my @requests = ApiManager::read_http_stream($t, 'bookstore.log');
+is(scalar @requests, 1, 'Backend received two requests');
+
+my $r = shift @requests;
+is($r->{verb}, 'GET', 'Backend request was a get');
+is($r->{uri}, '/shelves?key=this-is-an-api-key', 'Backend uri was /shelves');
+is($r->{headers}->{host}, "127.0.0.1:${BackendPort}", 'Host header was set');
+
+@requests = ApiManager::read_http_stream($t, 'servicecontrol.log');
+is(scalar @requests, 2, 'Service control received one request');
```

Comment:
The report should be called even for rejected request.  why it is not called.

Reply:
Turned off report cache